core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -91,7 +90,12 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
   /** List all the files currently stored on disk by the disk manager. */
   def getAllFiles(): Seq[File] = {
     // Get all the files inside the array of array of directories
-    subDirs.flatten.filter(_ != null).flatMap { dir =>
+    subDirs.flatMap { dir =>
```

Comment:
This is simply a matter of program order, not multi-threading. You can't see `subDirs(dirId)(subDirId) = newDir` happen before `newDir.mkdir()` no matter what. Above, you are fixing a simple race, that `subDirs(dirId)(subDirId)` can be observed `null` and then `if (subDir == null) {` is entered just before another thread sets the same to a value.